lib/event-provider/index.js

```diff
@@ -0,0 +1,141 @@
+    constructor(nahmiiDomainOrProvider) {
+        if (!nahmiiDomainOrProvider)
+            throw new TypeError('Argument should be string or NahmiiProvider');
+
+        let domain = nahmiiDomainOrProvider;
+        if (nahmiiDomainOrProvider instanceof NahmiiProvider) {
+            _provider.set(this, nahmiiDomainOrProvider);
+            domain = nahmiiDomainOrProvider.nahmiiDomain;
+        }
+
+        if (typeof domain !== 'string')
+            throw new TypeError('Argument should be string or NahmiiProvider');
+
+        _eventEmitter.set(this, new EventEmitter());
+        _socket.set(this, io.connect(`https://${domain}/receipts`, {path: '/events/socket.io'}));
+        subscribeToEvents.call(this);
+    }
+
+    /**
+     * Creates a new NahmiiEventProvider based on the input provided.
+     * @param {string|NahmiiProvider} nahmiiDomainOrProvider - The nahmii domain to connect to or an already connected provider.
+     * @returns {Promise<NahmiiEventProvider>}
+     */
+    static async from(nahmiiDomainOrProvider) {
+        if (!nahmiiDomainOrProvider)
+            return null;
+
+        if (nahmiiDomainOrProvider instanceof NahmiiProvider)
+            return new NahmiiEventProvider(nahmiiDomainOrProvider);
+
+        if (typeof nahmiiDomainOrProvider !== 'string')
+            return null;
+
+        return new NahmiiEventProvider(nahmiiDomainOrProvider);
+    }
+
+    /**
+     * Registers listener function so it will be called whenever a new Receipt
+     * has been issued by the nahmii cluster.
+     * Multiple calls passing the same listener function will result in the
+     * listener being added, and called, multiple times.
+     * Returns a reference to the NahmiiEventProvider, so that calls can be
+     * chained.
+     * @param listener
+     * @returns {NahmiiEventProvider}
+     */
+    onNewReceipt(listener) {
+        _eventEmitter.get(this).on(EventNames.newReceipt, receipt => {
+            listener.call(this, receipt);
+        });
+        return this;
+    }
+}
+
+/**
+ * Subscribes to critical socket.io events and relays nahmii events to any
+ * registered listeners.
+ * Private method, invoke with 'this' bound to provider instance.
+ * @private
+ */
+function subscribeToEvents() {
+    const socket = _socket.get(this);
+
+    function onEventApiError(error) {
+        dbg('Event API connection error: ' + JSON.stringify(error));
+    }
+
+    socket.on('pong', (latency) => {
+        dbg(`Event API latency: ${latency} ms`);
+    });
+    socket.on('connect_error', onEventApiError);
+    socket.on('error', onEventApiError);
+    socket.on('disconnect', onEventApiError);
+    socket.on('reconnect_error', onEventApiError);
+    socket.on('reconnect_failed', () => {
+        onEventApiError('Reconnecting to the Event API failed');
+    });
+
+    socket.on('new_receipt', receiptJSON => {
+        const receipt = Receipt.from(receiptJSON);
+        _eventEmitter.get(this).emit(EventNames.newReceipt, receipt);
```

Comment:
It's interesting that you can use a symbol when emitting events. I was not aware of that!

Reply:
You can use anything that will be coerced to a string AFAIK. :)